File: backend/app/models/analysis.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime

class IngredientInput(BaseModel):
    name: str

class AnalysisRequest(BaseModel):
    ingredients: List[str]
    session_id: Optional[str] = None

class AnalysisResponse(BaseModel):
    session_id: str
    safety_analysis: str
    research_attempts: int
    analysis_attempts: int
    critic_approved: bool
    research_confidence: float
    qdrant_hits: int
    tavily_hits: int
    total_critic_rejections: int
    analysis_start_time: datetime
    analysis_end_time: datetime
    
    class Config:
        from_attributes = True

class AnalysisProgress(BaseModel):
    stage: str  # "research", "analysis", "critic", "complete"
    progress: int  # 0-100
    message: str

File: frontend/src/types/analysis.ts
export interface AnalysisRequest {
  ingredients: string[];
  session_id?: string;
}

export interface AnalysisResponse {
  session_id: string;
  safety_analysis: string;
  research_attempts: number;
  analysis_attempts: number;
  critic_approved: boolean;
  research_confidence: number;
  qdrant_hits: number;
  tavily_hits: number;
  total_critic_rejections: number;
  analysis_start_time: string;
  analysis_end_time: string;
}

export interface AnalysisProgress {
  stage: string;
  progress: number;
  message: string;
}

export interface HistoryEntry {
  timestamp: string;
  ingredient_names: string[];
  safety_analysis?: string;
  critic_approved: boolean;
  research_confidence: number;
}
